Add Heraldicon integration

or opened the pull request:
For https://github.com/mirnovov/heraldtron/issues/1

Apologies for the noise, I noticed there was a lot of trailing whitespace, I thought it prudent to get that out of the way to avoid conflicts or noise in later commits.

Also noticed a trailing comma in the schema.

The final commit should be easier to review on its own, tho.

or committed:
Remove trailing whitespace, ensure final newline

or committed:
Fix schema

or committed:
Add Heraldicon blazonry integration

The !hd command allows parsing and rendering blazonry via
heraldicon.org, display the resulting PNG, and provide a link to where
the parsed arms can be edited directly.

In the case of errors the confusing string is indicated and
suggestions are displayed.

The !hd-options command displays all options that can be put as
keywords in front of the blazon with the !hd command.

Details: https://github.com/mirnovov/heraldtron/issues/1

mirnovov commented on ht/services.py:
```diff
@@ -38,39 +38,126 @@ def image_result(item):
+							"blazon": blazon,
+							"options": options,
+						},
+					})
```
```suggestion
	result = await utils.post_json(session, "https://2f1yb829vl.execute-api.eu-central-1.amazonaws.com/api",
		{
			"call": "generate-from-blazon",
			"data": {
				"blazon": blazon,
				"options": options,
			},
		}
	)
```
More in line with my standard code style

mirnovov commented on ht/services.py:
```diff
@@ -38,39 +38,126 @@ def image_result(item):
+		embed.set_thumbnail(url = "https://cdn.heraldicon.org/img/heraldicon-logo.png")
+		embed.set_image(url = png_url)
+		embed.add_field(name = "blazon", value = f"*{blazon}*", inline = True)
+		embed.set_footer(text = "CC BY-SA 4.0, attribution on https://heraldicon.org")
```
```suggestion
		png_url = success["png-url"]
		edit_link = f"[View in Heraldicon]({success['edit-url']})"
		embed = embeds.DRAW.create("", edit_link, heading = "Shield created!")
		embed.set_image(url = png_url)
		embed.add_field(name = "Blazon", value = f"*{blazon}*", inline = True)
		embed.set_footer(
			icon_url = "https://cdn.heraldicon.org/img/heraldicon-logo.png",
			text = "Drawn using Heraldicon; licensed under CC BY-SA 4.0 (attribution on https://heraldicon.org)"
		)
```
A few changes here:
- "Success" is kinda programmer-y lol, this is more consistent with the DS command
- I understand that you want to advertise your website, but the icon on the side is simply too big imo and kinda distracting if someone's doing a whole bunch of commands. I've moved it to the footer.
- Replaced the raw url with a link, looks a bit nicer imo

mirnovov commented on ht/services.py:
```diff
@@ -38,39 +38,126 @@ def image_result(item):
+					{
+						"call": "blazon-options",
+					})
+
```
```suggestion
	result = await utils.post_json(
		session, 
		"https://2f1yb829vl.execute-api.eu-central-1.amazonaws.com/api",
		{"call": "blazon-options"}
	)
```
again, nicer formatting imo

mirnovov commented on ht/services.py:
```diff
@@ -38,39 +38,126 @@ def image_result(item):
+	embed = embeds.DRAW.create("", "Unknown error.", heading = "Error")
+	embed.set_thumbnail(url = "https://cdn.heraldicon.org/img/heraldicon-logo.png")
+
+	return embed
```
```suggestion
		embed = embeds.DRAW.create(
			"", 
			"Put any of these options as first words before the blazon.", 
			heading = "Heraldicon rendering options"
		)
				
		add_option_type(embed, result["success"]["options"]["miscellaneous"], "General")
		add_option_type(embed, result["success"]["options"]["mode"], "Mode")
		add_option_type(embed, result["success"]["options"]["escutcheon"], "Escutcheons")
		add_option_type(embed, result["success"]["options"]["theme"], "Themes")
		add_option_type(embed, result["success"]["options"]["texture"], "Textures")
		
		embed.set_footer(
			icon_url = "https://cdn.heraldicon.org/img/heraldicon-logo.png",
			text = "For more info, see https://heraldicon.org"
		)
		return embed

	raise utils.CustomCommandError(
		"Cannot retrieve Heraldicon options",
		"An unknown error occured while trying to retrieve the Heraldicon options."
	)
	
def add_option_type(embed, source, name):
	embed.add_field(
		name = name, 
		value = ", ".join(f"`:{x}`" for x in source), 
		inline = False
	)
```

- Killed repeat code by moving the embed fields into a function (this is done elsewhere in the bot as well)
- The bot uses a standard error mechanism throughout the code, so replaced your error embed with that
- Again, relegate the icon to the footer

mirnovov commented:
Anyway, thanks very much for actually getting your hands dirty and adding the code haha, hopefully my changes aren't too much of an ask

or commented:
Groovy!
- I noticed a little bug with the error display, fixed that.
- Changed link text to "Edit on Heraldicon".
- Moved the `add_option_type` above `heraldicon_options`, so it is defined before.
- Fixed a spelling mistake.

The icon in the footer is a good idea, I didn't see that that can be done. I mainly wanted it to distinguish the responses from the DS ones, but it's fine down there as well.

mirnovov commented:
Thanks very much!
